### PR TITLE
Add `@wordpress/menus` package

### LINKED ISSUE #31580: Add @wordpress/menus package to better share code between nav block and nav editor
## What problem does this address?

Currently the **[Navigation _block_](https://github.com/WordPress/gutenberg/tree/trunk/packages/block-library/src/navigation)** and the **[Navigation _editor_ screen](https://github.com/WordPress/gutenberg/tree/trunk/packages/edit-navigation)** are separate codebases in separate packages. 

In one way this makes sense as it follows the established pattern in Gutenberg. Yet given that both features share a **similar set of task requirements** (eg: handling classic menus, converting to blocks...etc), I have observed **an increasing amount of duplication** between the two code bases.

This is causing problems for 

* maintenance.
* consistency between editor and block.
* overhead/friction for the development of new features.

To mitigate these problems, I am proposing the creation of a `@wordpress/menus` package to share code between Navigation editor and block.

But first, allow me to explain why this is necessary.

### Why is duplication a problem?

[As @gwwar suggests](https://github.com/WordPress/gutenberg/pull/31586#issuecomment-836890262), code duplication is generally a good first step to avoid tight coupling when we're still exploring a problem space. 

However, over time as these duplications grow in quantity and complexity it becomes increasingly difficult to avoid drift. This very quickly leads to bugs and regressions being easily introduced. For example, new contributors come in to fix bug but don't realize that duplicate code exists in other places (indeed even seasoned contributors are capable of doing this!).

Placing too much reliance on humans to keep track of a large amount of duplicate code is a recipe for an unstable codebase. Moreover, it also places an additional burden on developers when authoring and/or reviewing code as this must now be performed in several locations thereby increasing the complexity of any given task and leading to unnecessary cognitive load. Again this leads to mistakes being made.

### What examples do we have of these problems occurring?

Here is [a basic example of a **fix** for a regression that occurred when a refactoring effort on the Navigation block was not duplicated into the Navigation editor](https://github.com/WordPress/gutenberg/pull/31602). Both the author and the reviewer of the PR which originally introduced the regression are very familiar with the block and editor yet both still managed to forget about the need to duplicate the code. 

In this instance, the knock-on effect was minor ([Post Tag menu items were not being converted into their block equivalents in the Navigation block](https://github.com/WordPress/gutenberg/pull/31602/files#diff-57ec0477fdb7d78c0e2413a427cc94ad0c1e697a82475cfd07aa3b92fd3b49dbR51)) but a similar error could easily have more impactful consequences.

You will also note the same PR also corrected other minor inconsistencies.

This is just a single example of many such regressions and one which I believe illustrates the type of problems that occur when complex and rapidly evolving code is duplicated. 

### What duplication do we currently have?

Let's review the key and most impactful areas where the current duplication exists:

- Similar (but yet _unintentionally_ divergent) code which converts menu items to `core/navigation-link` blocks. Examples:
  - [Editor](https://github.com/WordPress/gutenberg/blob/d37bcb48d7028aa7701ff129ad6e338c65d1666f/packages/edit-navigation/src/store/resolvers.js#L95)
  - [Block](https://github.com/WordPress/gutenberg/blob/d37bcb48d7028aa7701ff129ad6e338c65d1666f/packages/block-library/src/navigation/placeholder.js#L38).
- Code for fetching Menus and Menu Items. Examples:
  - [Block](https://github.com/WordPress/gutenberg/blob/d37bcb48d7028aa7701ff129ad6e338c65d1666f/packages/block-library/src/navigation/placeholder.js#L52-L119).
  - [Editor](https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-navigation/src/hooks/use-navigation-editor.js).

The example above are both relatively complex and involved pieces of code. In addition, they both require the developer to maintain a (in brain) memory map of the relationship between `nav_menu_item` and `block.attributes` which is a significant cognative load. 

I see both as ripe for the introduction of regressions as evidenced by the existing drift in the two codebases for similar functional requirements.

## Why do we need this now? 

It's worth asking whether it would be best simply to bring both implementations into sync and then make more effort to maintain the duplication by hand.

However, I would argue that the requirements and complexity of both features have already become significant and the overhead to maintaining these "by hand" is close to becoming (if not already) overwhelming.

As new requirements are placed upon both block and editor this situation will inevitably become even worse.

### Aside: placeholder should remain in the Block
We are now  [looking to introduce a specific placeholder state for the Nav editor](https://github.com/WordPress/gutenberg/issues/23953). This is going to introduce yet more complexity and was the initial inpetus for my proposal to introduce the `@wordpress/menus` package.

Some have argued (with some validity I might add) that [the placeholder state should remain in the block and not be part of the editor](https://github.com/WordPress/gutenberg/issues/31580#issuecomment-836795478). 

Whilst this is indeed a valid point, and one I used to buttress my original proposal, I now see it as _tangential_ to the requirement to introduce a shared `@wordpress/menus` package in order to reduce complexity in the Nav editor and block.

The complexity is _already_ significant enough to merit finding a way to share code more effectively.

## What is your proposed solution?

As you're aware by now, I'm proposing that we create a `@wordpress/menus` package (naming of this is up for debate - I'm not fixed on `menus`) where we can colocate code that is shared between block and editor. This will reduce maintenance overhead, reduce cognitive load on developers and improve the resilience of the two features.

It will also provide the additional benefit of allowing us to more easily isolate and test some of the (duplicate) code which is not currently covered by automated tests.

### Existing precedent and whether to create co-dependent packages

I would draw your attention to the [`@wordpress/widgets` package](https://github.com/WordPress/gutenberg/tree/trunk/packages/widgets) which was created to share code between Widgets and Customizer screens. 

Whilst I'm aware that not everyone is in favour of such packages, it seems that the requirements of Gutenberg's development are necessitating some means of sharing code between packages. The discussion about _how_ we do this feels like an architectural decision that is rapidly becoming something that needs to be addressed. 

In the interests of moving forward in this particular instance however, I would argue that the existing precedent of the `@wordpress/widgets` package justifies the creation of a similar package for Navigation. 

I appreciate that @mtias and @youknowriad may feel that in order to avoid compounding a pattern they are not in favour of we first need to solve the wider problem of code sharing reuse in packages. I would caution, however, that whilst we await the development of such a solution, the Navigation components will inevitably be impacted by the friction (as illustrated above) generated by excessive code duplication between the Navigation block and editor.

## Conclusion

In conclusion, for the reasons outlined above, I recommend we proceed with the introduction of a shared `@wordpress/menus` package. This will:

* reduce complexity.
* reduce cognitive load on contributors and maintainers.
* reduce the likelihood of regressions across block and editor.
* reduce the likelihood of _inconsistencies_ between block and editor.
* improve our test coverage for key logic and functionality.

I would draw your attention to [the existing discussion on the PR](https://github.com/WordPress/gutenberg/pull/31586#issuecomment-834293572) where many of the points illustrated above were first introduced (particularly [this one](https://github.com/WordPress/gutenberg/pull/31586#issuecomment-836890262) by @gwwar). 

I look forward to us all producing a more consistent and reliable experience across Nav block and editor.

cc @gwwar @jasmussen @talldan @draganescu @Mamaduka

### PR DESCRIPTION
**Pull request by getdave**
## Description
Introduces a `@wordpress/menus` package as a place to store shared components in preparation for adding a unique placeholder state to the Navigation editor screen. 

As this placeholder will require much of the same logic, data fetching and mechanics as the block placeholder it makes sense to centralise what we can.

Closes https://github.com/WordPress/gutenberg/issues/31580

## How has this been tested?

* Nothing should change in either the Nav editor or the Nav block.
* All unit and e2e tests should continue to pass "as is".

## Types of changes
New feature (non-breaking change which adds functionality) 

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by getdave** [comment]
I feel this PR now has enough changes to demonstrate the utility of this package. I won't take things further (although I could) based on the view that we should merge early and then continue to iterate.

**Comment by gziolo** [comment]
> I feel this PR now has enough changes to demonstrate the utility of this package. I won't take things further (although I could) based on the view that we should merge early and then continue to iterate.

It's very important to have a good plan for what this new package would include. All constants, methods, and components exposed from this package will become public API for the WordPress core that we need to support indefinitely. They will be exposed under `wp.menus` global and all plugins can use them.

**Comment by youknowriad** [comment]
Also, is this a package that can be considered WP Agnostic (most packages should be ideally). I personally don't like packages that are just there to separate things and hold a bunch of reusable stuff, a package should have a clear purpose and can be used on its own do to something meaningful.

**Comment by getdave** [comment]
@youknowriad It is indeed separating things and holding reusable code.

> I personally don't like packages that are just there to separate things and hold a bunch of reusable stuff

Given the code duplication and potential for regression having lots of duplicate code introduces (not always but in this case I believe it does/has already) what other avenues does the Gutenberg project provide allow us to share code between packages? 

We already have this pattern in play in:

* `@wordpress/media-utils` - even the name gives this one away
* `@wordpress/widgets`

As such it feels like there is a need for a means of code reuse across modules. If packages are the way forward then it appears there is a argument for introducing a different mechanism - perhaps one that isn't published to the registry?

**Comment by getdave** [comment]
FYI this Issue https://github.com/WordPress/gutenberg/issues/31599 highlights one the main use cases for needing to refactor.

**Comment by youknowriad** [comment]
I think sharing things in small modules is fine, we just need the module's purpose to be good. Also we should resist as much as possible creating modules to share small things. I think duplicating is actually good compared to the introduction of bad abstractions we could regret later. So if it's just a couple small functions, I'm very fine duplicating. 

> We already have this pattern in play in:
> @wordpress/media-utils - even the name gives this one away
> @wordpress/widgets
> @wordpress/reusable-blocks

To be honest, I'm not a big fan of these packages. 

--
All that said, if If I'm not a fan of this kind of packages, I've came to "accept" their existence sometimes as the shared code becomes big.

File: packages/edit-navigation/src/store/utils.js [hunk 1]
```diff
@@ -7,6 +7,7 @@ import { keyBy, omit } from 'lodash';
  * WordPress dependencies
  */
 import { serialize } from '@wordpress/blocks';
+import { NEW_TAB_TARGET_ATTRIBUTE } from '@wordpress/menus';
```
Comment (talldan) [inline packages/edit-navigation/src/store/utils.js]: This constant should probably belong in the DOM package, but It might be better for it not to be a package export.

Either duplicating it or moving `blockAttributesToMenuItem` to the menus package would work.

**Review by talldan — COMMENTED** [review]
May not be useful, but I just remembered the other idea I had for refactoring this code, which was to treat the block <--> menu item conversion as a block transform with a 'menuItem' type (or perhaps a way to describe an `entity` to `block` transform? Page to Nav Link may also be useful).

There are a few aspects that I was unclear on, so I hadn't pushed on with the idea:
- Is this too specific for the Navigation Link block? It might end up being the only block that has this transform. Should third party blocks be able to add such a transform type?
- How could this be used for back compat/extensibility? (e.g. extending the conversion process to allow converting custom menu item fields to block attributes, which might relate to https://github.com/WordPress/gutenberg/issues/31551). At the moment I don't know enough about how menu item custom fields work—needs more investigation.

**Comment by getdave** [comment]
> I think sharing things in small modules is fine, we just need the module's purpose to be good

The primary purpose would be to share the code which handles fetching and converting menus in a central location.

I agree duplication is sometimes ok, but this code is "involved" (not in terms of complexity but in terms of remembering how blocks map to menus and vice versa) and I've already encountered several regressions where the cause was the duplicate code in the editor and the block being out of sync (here's [another example](https://github.com/WordPress/gutenberg/pull/31602)).

I'm not sure it's practical or reliable to rely on humans to keep this amount of code in sync. 

Would it help if I went and laid out a very thorough proposal for this package which a lot of detail about the specifics of how and why we need this package? 

In the meantime, I'm going to continue to manually bring the duplicate code into line between block and editor. That way if we do get a green light on this package it'll be a lot easier.

**Comment by youknowriad** [comment]
> I agree duplication is sometimes ok, but this code is "involved" (not in terms of complexity but in terms of remembering how blocks map to menus and vice versa) and I've already encountered several regressions where the cause was the duplicate code in the editor and the block being out of sync (here's another example).

Shared complex code sounds like a good argument, I just to want to make sure the maintainability costs are considered :) 

> Would it help if I went and laid out a very thorough proposal for this package which a lot of detail about the specifics of how and why we need this package?

I think you can just update the README me here to reflect that right?

File: packages/menus/src/index.js [hunk 2]
```diff
@@ -0,0 +1,2 @@
+export * from './utils';
```
Comment (gwwar) [inline packages/menus/src/index.js]: Minor note: I tend to think of utils as a junk drawer for naming. Since we have a whole new package here, what might be more descriptive to readers and a better fit? 

What the logic currently does is related to transformation/serialization. What other type of things might we add to this package?

**Comment by gwwar** [comment]
Leaving a couple of high level thoughts here:

#### Code Duplication leads to drift

Code duplication is generally a good first step to avoid tight coupling when we're still exploring a problem space. However this also means that implementations **_will_** drift over time, regardless of how diligent folks might be. This can happen very easily, for example new contributors come in to fix bug but don't realize that this code exists in many other places.

In some cases the drift is fine since we realize that use cases will differ, but if that isn't the case and we'd like to keep logic in sync, some shared place to store the code is necessary.

#### Many Small Packages (per feature) vs One Shared Package?
 
While `@wordpress/media-utils` and `@wordpress/widgets` already exist, are we sure we like this pattern before we add another package? I think both options can work and have different tradeoffs, I just want us to make sure we think through this before committing to the pattern.

It's likely that there will be other new features that need to share common logic or default implementations across our block editors (site / post / widget / navigation / any new ones we can build). There was a bit of exploration in https://github.com/WordPress/gutenberg/pull/29993 for this, but a single experimental function wasn't a great candidate.

#### Naming

If we do choose to pick many small packages, does Menus work for us? If I recall correctly the Navigation block was named Navigation over Menu to avoid confusion over restaurant menus.

#### __experimentalFetchLinkSuggestions

This currently resides in `@wordpress/core-data` as a temporary holding place. If we centralize navigation/menu transform logic, I'd be happy to move it out of `wordpress/core-data`.

**Comment by getdave** [comment]
Thanks, everyone for your valuable input. Somehow this PR became the place to discuss the merit of the feature. 

My fault. 

Let's move further discussion to [the Issue](https://github.com/WordPress/gutenberg/issues/31580) where [I will summarize thoughts and respond to Matias's question](https://github.com/WordPress/gutenberg/issues/31580#issuecomment-838147856).

File: packages/menus/src/index.js (hunk 2 above)
Comment (getdave) [reply packages/menus/src/index.js]: Removed.

File: packages/edit-navigation/src/store/utils.js (hunk 1 above)
Comment (getdave) [reply packages/edit-navigation/src/store/utils.js]: Removed

**Comment by getdave** [comment]
We're _probably_ not going to pursue this at this point. If we need to we can revisit. Closing for now.
